metric-publishers/cloudwatch-metric-publisher/src/main/java/software/amazon/awssdk/metrics/publishers/cloudwatch/internal/task/UploadMetricsTasks.java

```diff
@@ -0,0 +1,65 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.metrics.publishers.cloudwatch.internal.task;
+
+import static software.amazon.awssdk.metrics.publishers.cloudwatch.internal.CloudWatchMetricLogger.METRIC_LOGGER;
+
+import java.util.Collections;
+import java.util.List;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.metrics.publishers.cloudwatch.CloudWatchMetricPublisher;
+import software.amazon.awssdk.metrics.publishers.cloudwatch.internal.MetricUploader;
+import software.amazon.awssdk.metrics.publishers.cloudwatch.internal.transform.MetricCollectionAggregator;
+import software.amazon.awssdk.services.cloudwatch.model.PutMetricDataRequest;
+
+/**
+ * A task that is executed on the {@link CloudWatchMetricPublisher}'s executor to collect requests from a
+ * {@link MetricCollectionAggregator} and write them to a {@link MetricUploader}.
+ */
+@SdkInternalApi
+public class UploadMetricsTasks implements Runnable {
+    private final MetricCollectionAggregator collectionAggregator;
+    private final MetricUploader uploader;
+    private int maximumRequestsPerFlush;
+
+    public UploadMetricsTasks(MetricCollectionAggregator collectionAggregator,
+                              MetricUploader uploader,
+                              int maximumRequestsPerFlush) {
+        this.collectionAggregator = collectionAggregator;
+        this.uploader = uploader;
+        this.maximumRequestsPerFlush = maximumRequestsPerFlush;
+    }
+
+    @Override
+    public void run() {
+        List<PutMetricDataRequest> allRequests = collectionAggregator.getRequests();
+        List<PutMetricDataRequest> requests = allRequests;
+        if (requests.size() > maximumRequestsPerFlush) {
+            METRIC_LOGGER.warn(() -> "Maximum AWS SDK client-side metric call count exceeded: " + allRequests.size() +
+                                     " > " + maximumRequestsPerFlush + ". Some metric requests will be dropped. This occurs when "
+                                     + "the caller has configured too many metrics or too unique of dimensions without an "
+                                     + "associated increase in the maximum-calls-per-upload configured on the publisher.");
+
+            // Randomly pick which requests we intend to keep/drop so that we aren't always dropping the same ones if this happens
+            // a lot of times in a row.
+            Collections.shuffle(requests);
```

Comment:
I don't think that's what we concluded, but to avoid arguing, I'll just remove it.

Reply:
I think our options are 1) send nothing  2) send as much as we can but undetermined order  3) send as much as we can but determined order   4) send as much as we can but actively shuffle. I think no matter what option we pick the settings need to be changed as this isn't a good situation for the customer. I'm almost leaning towards 1) under the assumption that no data might be better than incomplete/misleading data especially where aggregates are concerned but 2 might be more appropriate because it wouldn't fail all the time necessarily so there may be partial data anyway. Doing 3 or 4 seems like we're saying this is another kind of happy path and I don't think I'm ready to agree with that at this point.

Reply:
Crazy idea but can we buffer and try and send them in subsequent requests? Obviously the buffer can fill up at which point we really have to drop stuff on the ground, but it would give the system a bit of elasticity to deal with spikes.

Reply:
(3) is also quite difficult to implement. (1) could still result in inconsistent values, since previous and future flushes could include data from the 'time bucket' of the dropped values. I think (2) or (4) are both probably fine, from the customer perspective. In that case (2)  might win since it's less code.